modules/component/src/listComponentState.ts

```diff
@@ -0,0 +1,33 @@
+import {List} from 'standard-data-structures'
+
+type keyType = string | number
+type ListItem<S> = {state: S; key: keyType}
+export class ListComponentState<S> {
+  private constructor(
+    private baseInit: () => S,
+    private readonly items: List<ListItem<S>> = List.empty(),
+    private readonly lookUp: {[k in keyType]?: List<ListItem<S>>} = {}
+  ) {}
+  static of<T>(baseInit: () => T) {
+    return new ListComponentState(baseInit)
+  }
+  insertItem(k: keyType, value: S = this.baseInit()) {
+    const items = this.items.prepend({state: value, key: k})
+    return new ListComponentState(this.baseInit, items, {
+      ...this.lookUp,
+      [k]: items
+    })
+  }
+  getItem(k: keyType): S | null {
```

Comment:
rename to `get(k: KeyType): Option<S>` 

Option is available inside `standard-data-structures`